### ISSUE TITLE
Storing Embeddings in Lhotse

### ISSUE THREAD

**Issue by danpovey** [lhotse-speech/lhotse#453]
This is kind of a major feature request, IDK whether you have time for this, Piotr, but it's part of a big direction that I want to go in, in Icefall, and it is a good way to leverage the strengths of Lhotse.

It's related to things like BERT, HUBERT, wav2vec2, and so on... basically the idea is that we will be predicting some kind of code generated from an embedding that was obtained from the activations of a previously trained network.  This could be used either as the primary loss or as an auxiliary loss.  In general, I want to have a flexible setup where you have a place on disk where you store a bunch of embeddings produced by previous models, and you train your model to predict a bunch of them simultaneously.  The advantage of doing this via codes dumped to disk, and not computing them as needed, is that we don't have to worry about loading a bunch of different models into memory and having all their associated code still working.  Also it's more efficient compute-wise as we only have to dump the data once.

Right now, for ease of implementation, I am assuming that each embedding, on each frame, is going to be converted into a short sequence of labels, with the labels spanning 0..255 (1 byte), where each label likely represents some subset of dimensions of the embedding.  In practice I anticipate that we might have anywhere between 1 and 16 labels per frame.  So the natural way these would be stored for a single cut is as a byte array, of shape (num_frames, seq_len), where e.g. seq_len=4.  [For cases where seq_len is very small, we might want to consider just storing this inline in the jsonl file, rather than using hdf5, but this is a side issue.]

There are two types of workflow that we'd have, when dealing with these labels.

 (i) Dumping the labels.  The scenario is we have trained some kind of model, and we want to extract embeddings from it.  The part of this that interfaces with Lhotse, is that we want to dump these embeddings to disk, and the relevant metadata.   The modeling code will give us the int8 array, e.g. of shape (B, T, S) i.e. (batch-size, num-frames, bytes-per-frame).  [I'd like to give these codes a special name, "encoding"].   I suppose we could dump the encodings for each supervision segment, but in general I'd like to think of this as corresponding to cuts rather than supervision segments.   We can if needed assume here that there is no Musan or SpecAug augmentation and no cut concatenation (but there may be speed augmentation).  It might be kind of slow to have a single process dump the encodings for large datasets (think Gigaspeech XL, which is 20K hours or something like that).  It would be nice if there were a way to split and re-combine data, that would work with this.

 (ii) Loading the labels.  This is a little like loading supervisions.  Depending on the scenario, we might be loading the encodings AND supervisions, or just loading encodings.  Here is one possible scenario: we want to load the encodings, where available, for a list of 3 previously trained models.  These encodings have been written via manifest files as e.g. enc_train_clean_360_a3f3cd91.manifest.gz, enc_train_clean_360_d673afg52.manifest.gz, and so on (I'm assuming the suffix represents an md5 hash of the model, so we don't get models mixed up).  So the user wants a tensor of int8, of shape (B, T, 3, bytes-per-frame), where the user specifies the bytes-per-frame, together with a tensor of bool of the same shape that acts as a mask.  The tensor of int8 has zeros in masked positions.  Also, this bytes-per-frame may in general be different from the ones on disk, so we can mix and match; we'd pad with 0 or truncate as necessary to achieve this.  And the T is not going to be the same T as the features, since there is subsampling.  It is not essential for there to be an exact relationship between the T's; the modeling code can truncate or pad needed to deal with this.

I don't want to hardcode it that these are byte arrays, as we might in future want to use the same code for another purpose, but they are anticipated to be byte arrays for the time being.  For timing purposes I think we can just assume that frame 0 corresponds to the start of the cut, the user will provide the subsampling factor (e.g. 4), and the number of frames is proportional to the duration, e.g. rounded to the nearest int.  Other than that I don't think we need to be super exact about the numbers of frames; if there are weird effects at the boundaries I think the prediction model will just learn to deal with it.

I'm not sure how to deal with issues relating to how the cuts are sorted- I assume we have to sort by duration for purposes of batching, whereas the manifests might often be sorted a different way?

**Comment by pzelasko** [comment]
Yeah, this is interesting and I'd like to support that. 

For dumping the labels, we can re-use `cut_set.compure_and_store_features[_batch]` and wrap the model you want to use in a `FeatureExtractor`. I may add a function wrapper to streamline this for PyTorch models, or maybe just write some examples.

For storing the meta-data and interacting with it, I agree this should be done on Cut-level rather than Supervision-level. This is actually related to https://github.com/lhotse-speech/lhotse/issues/393 -- it would allow us to store arbitrary extra information, and also make it work out-of-the-box with padding/truncating etc. since the temporal dimension would be marked there.

Give me some time to figure these things out. I might come up with a workable prototype in 1-2 weeks.

**Comment by pzelasko** [comment]
Potential issues I see is the mismatch between T's of different representations (3 of them: duration seconds, audio samples, and num frames were already giving me headaches). Not sure yet what's the right solution but we'll figure it out. Maybe being less strict about the consistency is an option (or necessity).

I didn't understand the issues related to sorting. We are currently sorting on duration but in principle we can sort on anything we want.

**Comment by danpovey** [comment]
Great!  In the meantime I'll try to do some of the modeling aspects, e.g. the code to get the encodings.
Yes, I think we don't have to be too strict with the embeddings' duration, we can rely on padding and truncation if there are mismatches.
Regarding sorting, it's not really a big issue, but I was concerned about how we'll deal with large datasets where there jsonl files because it's too big to fit into memory.  I suppose it must be possible to re-sort that file after dumping it, somehow.

**Comment by pzelasko** [comment]
With current Lhotse, as long as you don't want to do bucketing, using `cuts = CutSet.from_jsonl_lazy(path)` lets you iterate JSONL in sampler without reading full file into the memory. So it should be fine.

I was thinking about extending this to bucketing but I have a better idea. PyTorch guys are working on "DataPipes", which make it easier to write these sort of things. I have working drafts of this with Lhotse somewhere. Ideally closer to Q1 2022 I'd be able to push out a new version of dataloading based on this.

**Comment by pzelasko** [comment]
BTW for the lazy JSONL reading, shuffling is also done on-the-fly with a buffer of ~10k samples. But if you want a true shuffle it's a bash one-liner.

**Comment by danpovey** [comment]
Something else I should mention: because the codes are specific to the model and to the quantizer used, I think it would make sense to attach some kind of short hash so we don't get multiple types of codes being conflated.  I'm thinking if the hash is short we could perhaps store it in each item of the manifest, so that it would be possible to combine manifests and not have a danger of losing track of which model is which, e.g. if filenames are changed.

In future I am contemplating having setups where the model is not designed to predict a fixed number of encodings, but has an "open-set" approach, somehow involving the hashes, converted to vectors.  I haven't pinned down the details of this, e.g. how these hashes would be conveyed to the modeling code.  I am just mentioning it early in case it would affect any of the design.  Even if we have a "closed-set" approach, the hashes may still be useful for avoiding mixups.

**Comment by pzelasko** [comment]
See the following two PRs for a draft of the overall solution:
- manifests for arbitrary arrays https://github.com/lhotse-speech/lhotse/pull/458
- extending cuts with custom fields + support for reading arrays conveniently https://github.com/lhotse-speech/lhotse/pull/459

**Comment by pzelasko** [comment]
This feature + some effort into refreshing docs and examples for older stuff should make a nice v1.0 release.

**Comment by danpovey** [comment]
Piotr, just to clarify, you plan to take care of the batching and data-saving aspects of this, right?
Scenario is: we train a model conventionally; we then run inference without SpecAug to get activations
from some internal encoder layer, and quantize them to some number of bytes per frame, e.g. with this 
code:
https://github.com/danpovey/quantization
.. and at that time we save those encodings, indexed by utterance-id.  
The idea, then, is that we train some other model where 
it has an auxiliary loss that predicts those encodings.  We can choose whether we want to give it some weight
on all frames, or just put weight on the SpecAug-masked frames, like in Hubert.  [We'd probably use SpecAug in the 
training script directly, without time-warping, rather than using the SpecAug from Lhotse, to make it easier to get the SpecAug masking tensor.]

**Comment by pzelasko** [comment]
Yes, what I'm building should work well with this scenario. I'll actually use these ideas for building test cases, and once it's ready, I'll show you an example how to use Lhotse for this.

BTW for SpecAug in this scenario, maybe `TimeMasking` and `FrequencyMasking` from torchaudio are useful https://pytorch.org/audio/stable/transforms.html#frequencymasking

**Comment by pzelasko** [comment]
I'm closing this as it seems to be working and stable; LMK if there's anything else that'd be helpful.
